File: src/sarma_cli/context/compaction.py
"""Token-window aware context compaction.

This module owns the policy for deciding when to compact conversation history
and how to split raw tail messages from older messages that should become
structured memory. It deliberately does not depend on Session or Store.
"""

from __future__ import annotations

from collections.abc import Awaitable, Callable
from dataclasses import dataclass

from sarma_cli.engine.models import ConversationMessage

Summarizer = Callable[[list[ConversationMessage]], Awaitable[str]]


@dataclass(frozen=True, slots=True)
class ContextWindowPolicy:
    """Budget knobs for automatic and manual compaction."""

    max_context_tokens: int
    trigger_ratio: float = 0.90
    raw_tail_ratio: float = 0.55
    output_reserve_ratio: float = 0.12
    minimum_output_reserve_tokens: int = 2_048
    static_prompt_tokens: int = 0

    @property
    def budget(self) -> int:
        return max(int(self.max_context_tokens or 1), 1)

    @property
    def trigger_tokens(self) -> int:
        return max(int(self.budget * self.trigger_ratio), 1)

    @property
    def raw_tail_tokens(self) -> int:
        return max(int(self.budget * self.raw_tail_ratio), 1)

    @property
    def output_reserve_tokens(self) -> int:
        ratio_reserve = int(self.budget * self.output_reserve_ratio)
        return max(ratio_reserve, self.minimum_output_reserve_tokens)

    @property
    def fixed_overhead_tokens(self) -> int:
        return max(self.static_prompt_tokens, 0) + self.output_reserve_tokens


@dataclass(frozen=True, slots=True)
class CompactionPlan:
    """A deterministic split of history into structured memory and raw tail."""

    should_compact: bool
    keep_tail: list[ConversationMessage]
    older: list[ConversationMessage]
    estimated_input_tokens: int
    trigger_tokens: int


class ContextCompactor:
    """Apply context-window policy to conversation history."""

    def __init__(self, policy: ContextWindowPolicy) -> None:
        self._policy = policy

    @property
    def policy(self) -> ContextWindowPolicy:
        return self._policy

    def plan(
        self,
        history: list[ConversationMessage],
        *,
        upcoming_text: str = "",
        force: bool = False,
    ) -> CompactionPlan:
        estimated = (
            self.estimate_history_tokens(history)
            + self.estimate_text_tokens(upcoming_text)
            + self._policy.fixed_overhead_tokens
        )
        should_compact = force or estimated >= self._policy.trigger_tokens
        keep_tail: list[ConversationMessage] = []
        older: list[ConversationMessage] = []

        if should_compact:
            keep_tail, older = self.split_raw_tail(history)
            should_compact = bool(older)

        return CompactionPlan(
            should_compact=should_compact,
            keep_tail=keep_tail,
            older=older,
            estimated_input_tokens=estimated,
            trigger_tokens=self._policy.trigger_tokens,
        )

    async def compact(
        self,
        history: list[ConversationMessage],
        summarize: Summarizer,
        *,
        conversation_id: str = "",
        upcoming_text: str = "",
        force: bool = False,
    ) -> tuple[bool, list[ConversationMessage], str]:
        """Return ``(changed, new_history, memory_text)``."""
        plan = self.plan(history, upcoming_text=upcoming_text, force=force)
        if not plan.should_compact:
            return False, history, ""

        memory = (await summarize(plan.older)).strip()
        if not memory:
            return False, history, ""

        memory_message = ConversationMessage(
            conversation_id=conversation_id,
            turn_id="compact",
            role="system",
            content=build_memory_context_message(memory),
        )
        return True, [memory_message, *plan.keep_tail], memory

    def split_raw_tail(
        self,
        history: list[ConversationMessage],
    ) -> tuple[list[ConversationMessage], list[ConversationMessage]]:
        tail: list[ConversationMessage] = []
        total = 0
        target = self._policy.raw_tail_tokens

        for message in reversed(history):
            cost = self.estimate_message_tokens(message)
            if tail and total + cost > target:
                break
            tail.append(message)
            total += cost

        tail.reverse()
        older_count = len(history) - len(tail)
        if older_count == 0 and total > target:
            return [], list(history)
        return tail, history[:older_count]

    def estimate_history_tokens(self, messages: list[ConversationMessage]) -> int:
        return sum(self.estimate_message_tokens(message) for message in messages)

    @classmethod
    def estimate_message_tokens(cls, message: ConversationMessage) -> int:
        framing_tokens = 16
        role_tokens = cls.estimate_text_tokens(message.role)
        content_tokens = cls.estimate_text_tokens(message.content)
        reasoning_tokens = cls.estimate_text_tokens(message.reasoning_content or "")
        return framing_tokens + role_tokens + content_tokens + reasoning_tokens

    @staticmethod
    def estimate_text_tokens(text: str) -> int:
        # Provider-neutral fallback. Exact provider tokenizers can be added
        # behind this interface without changing Session.
        return max(0, len(text or "") // 4)


def estimate_static_prompt_tokens(system_prompt: str, tool_count: int = 0) -> int:
    # Tool schemas are not available here, so reserve a conservative per-tool
    # budget in addition to the actual system prompt text.
    return ContextCompactor.estimate_text_tokens(system_prompt) + max(tool_count, 0) * 128


def build_memory_context_message(memory: str) -> str:
    return (
        "Structured memory compacted from prior conversation. Use it as "
        "durable context; do not treat it as a user request.\n\n"
        f"{memory.strip()}"
    )


STRUCTURED_MEMORY_PROMPT = """\
Compact the prior conversation into structured durable memory.

Return exactly these sections. Preserve all user constraints, decisions,
verified facts, unresolved tasks, and useful artifacts. Prefer precise file
paths, function names, commands, URLs, configuration keys, and test outcomes.
Do not optimize for shortness at the cost of losing facts.

Goals:
- ...

Constraints:
- ...

Decisions:
- ...

Entities:
- Files:
- Functions / symbols:
- Addresses / offsets:
- Commands:
- URLs:
- Other identifiers:

Verified Facts:
- ...

Tool Results:
- ...

Open Tasks:
- ...

Risks / Unknowns:
- ...

Do not include hidden chain-of-thought or verbose transcripts. Return a concise
structured memory artifact.
"""

File: src/context/compaction.ts
/**
 * Token-window aware context compaction.
 *
 * Owns the policy for deciding when to compact conversation history and how to
 * split raw tail messages from older messages that become structured memory.
 * Deliberately does not depend on Session or Store.
 */

import { ConversationMessage } from "@/engine/models";
import type { TokenEstimator } from "@/context/tokenizer";

export type Summarizer = (messages: ConversationMessage[]) => Promise<string>;

export interface ContextWindowPolicyInit {
  maxContextTokens: number;
  triggerRatio?: number;
  rawTailRatio?: number;
  outputReserveRatio?: number;
  minimumOutputReserveTokens?: number;
  staticPromptTokens?: number;
  tokenEstimator?: TokenEstimator;
}

/** Budget knobs for automatic and manual compaction. */
export class ContextWindowPolicy {
  readonly maxContextTokens: number;
  readonly triggerRatio: number;
  readonly rawTailRatio: number;
  readonly outputReserveRatio: number;
  readonly minimumOutputReserveTokens: number;
  readonly staticPromptTokens: number;
  readonly tokenEstimator: TokenEstimator;

  constructor(init: ContextWindowPolicyInit) {
    this.maxContextTokens = init.maxContextTokens;
    this.triggerRatio = init.triggerRatio ?? 0.9;
    this.rawTailRatio = init.rawTailRatio ?? 0.55;
    this.outputReserveRatio = init.outputReserveRatio ?? 0.12;
    this.minimumOutputReserveTokens = init.minimumOutputReserveTokens ?? 2_048;
    this.staticPromptTokens = init.staticPromptTokens ?? 0;
    this.tokenEstimator = init.tokenEstimator ?? ContextCompactor.estimateTextTokens;
  }

  get budget(): number {
    return Math.max(Math.trunc(this.maxContextTokens || 1), 1);
  }
  get triggerTokens(): number {
    return Math.max(Math.trunc(this.budget * this.triggerRatio), 1);
  }
  get rawTailTokens(): number {
    return Math.max(Math.trunc(this.budget * this.rawTailRatio), 1);
  }
  get outputReserveTokens(): number {
    const ratioReserve = Math.trunc(this.budget * this.outputReserveRatio);
    return Math.max(ratioReserve, this.minimumOutputReserveTokens);
  }
  get fixedOverheadTokens(): number {
    return Math.max(this.staticPromptTokens, 0) + this.outputReserveTokens;
  }
}

/** A deterministic split of history into structured memory and raw tail. */
export interface CompactionPlan {
  shouldCompact: boolean;
  keepTail: ConversationMessage[];
  older: ConversationMessage[];
  estimatedInputTokens: number;
  triggerTokens: number;
}

/** Apply context-window policy to conversation history. */
export class ContextCompactor {
  constructor(private readonly _policy: ContextWindowPolicy) {}

  get policy(): ContextWindowPolicy {
    return this._policy;
  }

  plan(
    history: ConversationMessage[],
    options: { upcomingText?: string; force?: boolean } = {},
  ): CompactionPlan {
    const upcomingText = options.upcomingText ?? "";
    const force = options.force ?? false;
    const estimated =
      this.estimateHistoryTokens(history) +
      ContextCompactor.estimateTextTokens(upcomingText) +
      this._policy.fixedOverheadTokens;
    let shouldCompact = force || estimated >= this._policy.triggerTokens;
    let keepTail: ConversationMessage[] = [];
    let older: ConversationMessage[] = [];

    if (shouldCompact) {
      [keepTail, older] = this.splitRawTail(history);
      shouldCompact = older.length > 0;
    }

    return {
      shouldCompact,
      keepTail,
      older,
      estimatedInputTokens: estimated,
      triggerTokens: this._policy.triggerTokens,
    };
  }

  /** Return `[changed, newHistory, memoryText]`. */
  async compact(
    history: ConversationMessage[],
    summarize: Summarizer,
    options: { conversationId?: string; upcomingText?: string; force?: boolean } = {},
  ): Promise<[boolean, ConversationMessage[], string]> {
    const plan = this.plan(history, { upcomingText: options.upcomingText, force: options.force });
    if (!plan.shouldCompact) return [false, history, ""];

    const memory = (await summarize(plan.older)).trim();
    if (!memory) return [false, history, ""];

    const memoryMessage = new ConversationMessage({
      conversationId: options.conversationId ?? "",
      turnId: "compact",
      role: "system",
      content: buildMemoryContextMessage(memory),
    });
    return [true, [memoryMessage, ...plan.keepTail], memory];
  }

  splitRawTail(history: ConversationMessage[]): [ConversationMessage[], ConversationMessage[]] {
    const tail: ConversationMessage[] = [];
    let total = 0;
    const target = this._policy.rawTailTokens;

    for (let i = history.length - 1; i >= 0; i--) {
      const message = history[i]!;
      const cost = this.estimateMessageTokens(message);
      if (tail.length && total + cost > target) break;
      tail.push(message);
      total += cost;
    }

    tail.reverse();
    const olderCount = history.length - tail.length;
    if (olderCount === 0 && total > target) {
      return [[], [...history]];
    }
    return [tail, history.slice(0, olderCount)];
  }

  estimateHistoryTokens(messages: ConversationMessage[]): number {
    return messages.reduce((sum, m) => sum + this.estimateMessageTokens(m), 0);
  }

  estimateMessageTokens(message: ConversationMessage): number {
    return ContextCompactor.estimateMessageTokens(message, this._policy.tokenEstimator);
  }

  static estimateMessageTokens(message: ConversationMessage, estimateText: TokenEstimator = ContextCompactor.estimateTextTokens): number {
    const framingTokens = 16;
    const roleTokens = estimateText(message.role);
    const contentTokens = estimateText(message.content);
    const reasoningTokens = estimateText(message.reasoningContent || "");
    return framingTokens + roleTokens + contentTokens + reasoningTokens;
  }

  static estimateTextTokens(text: string): number {
    // Provider-neutral fallback. Exact provider tokenizers can be added behind
    // this interface without changing Session.
    return Math.max(0, Math.floor((text || "").length / 4));
  }
}

export function estimateStaticPromptTokens(
  systemPrompt: string,
  toolCount = 0,
  estimateText: TokenEstimator = ContextCompactor.estimateTextTokens,
): number {
  // Tool schemas are not available here, so reserve a conservative per-tool
  // budget in addition to the actual system prompt text.
  return estimateText(systemPrompt) + Math.max(toolCount, 0) * 128;
}

export function buildMemoryContextMessage(memory: string): string {
  return (
    "Structured memory compacted from prior conversation. Use it as " +
    "durable context; do not treat it as a user request.\n\n" +
    memory.trim()
  );
}

export const STRUCTURED_MEMORY_PROMPT = `Compact the prior conversation into structured durable memory.

Return exactly these sections. Preserve all user constraints, decisions,
verified facts, unresolved tasks, and useful artifacts. Prefer precise file
paths, function names, commands, URLs, configuration keys, and test outcomes.
Do not optimize for shortness at the cost of losing facts.

Goals:
- ...

Constraints:
- ...

Decisions:
- ...

Entities:
- Files:
- Functions / symbols:
- Addresses / offsets:
- Commands:
- URLs:
- Other identifiers:

Verified Facts:
- ...

Tool Results:
- ...

Open Tasks:
- ...

Risks / Unknowns:
- ...

Do not include hidden chain-of-thought or verbose transcripts. Return a concise
structured memory artifact.
`;
